File: code_snippets/java/src/main/java/usecases/eventprocessing/eventtransactions/UserFeed.java
package usecases.eventprocessing.eventtransactions;

import static usecases.eventprocessing.eventtransactions.utils.Stubs.*;

import dev.restate.sdk.ObjectContext;
import dev.restate.sdk.annotation.Handler;
import dev.restate.sdk.annotation.VirtualObject;
import dev.restate.sdk.endpoint.Endpoint;
import dev.restate.sdk.lambda.BaseRestateLambdaHandler;
import java.time.Duration;
import usecases.microservices.SubscriptionService;

// <start_here>
@VirtualObject
public class UserFeed {

  public record SocialMediaPost(String content, String metadata) {}

  // <mark_1>
  @Handler
  public void processPost(ObjectContext ctx, SocialMediaPost post) {
    // </mark_1>
    // <mark_5>
    String userId = ctx.key();
    // </mark_5>

    // <mark_3>
    String postId = ctx.run(String.class, () -> createPost(userId, post));
    // </mark_3>

    // <mark_4>
    // <mark_3>
    while (ctx.run(String.class, () -> getPostStatus(postId)).equals("PENDING")) {
      // </mark_3>
      // <mark_2>
      ctx.sleep(Duration.ofSeconds(5));
      // </mark_2>
    }
    // </mark_4>

    // <mark_3>
    ctx.run(() -> updateUserFeed(userId, postId));
    // </mark_3>
  }
}

// <mark_6>
class MyLambdaHandler extends BaseRestateLambdaHandler {
  @Override
  public void register(Endpoint.Builder builder) {
    builder.bind(new SubscriptionService());
  }
}
// </mark_6>
// <end_here>

File: code_snippets/go/usecases/eventprocessing/eventtransactions/userfeed.go
package main

import (
	"context"
	"github.com/aws/aws-lambda-go/lambda"
	restate "github.com/restatedev/sdk-go"
	"github.com/restatedev/sdk-go/server"
	"log"
	"log/slog"
	"os"
	"time"
)

type SocialMediaPost struct {
	Content  string `json:"content"`
	Metadata string `json:"metadata"`
}

// <start_here>
type UserFeed struct{}

// <mark_1>
func (UserFeed) ProcessPost(ctx restate.ObjectContext, post SocialMediaPost) error {
	// </mark_1>
	// <mark_5>
	var userId = restate.Key(ctx)
	// </mark_5>

	// <mark_3>
	postId, err := restate.Run(ctx, func(ctx restate.RunContext) (string, error) {
		return CreatePost(userId, post)
	})
	// </mark_3>
	if err != nil {
		return err
	}

	// <mark_4>
	for {
		// <mark_3>
		status, err := restate.Run(ctx, func(ctx restate.RunContext) (string, error) {
			return GetPostStatus(postId), nil
		})
		// </mark_3>
		if err != nil {
			return err
		}
		if status != PENDING {
			break
		}
		// <mark_2>
		if err = restate.Sleep(ctx, 5*time.Second); err != nil {
			// </mark_2>
			return err
		}
	}
	// </mark_4>

	// <mark_3>
	if _, err := restate.Run(ctx, func(ctx restate.RunContext) (restate.Void, error) {
		return restate.Void{}, UpdateUserFeed(userId, postId)
		// </mark_3>
	}); err != nil {
		return err
	}

	return nil
}

// <mark_6>
func main() {
	handler, err := server.NewRestate().
		Bind(restate.Reflect(UserFeed{})).
		Bidirectional(false).
		LambdaHandler()
	if err != nil {
		log.Fatal(err.Error())
	}
	lambda.Start(handler)
}

// </mark_6>
// <end_here>
